File: 2023/04.10/서지훈_백준_2749_피보나치수3.java
import java.io.*;
import java.math.BigInteger;

public class Main {
    public static long[][] matrix_multiply(long[][] mat1, long[][] mat2){
        return new long[][]{{(mat1[0][0] * mat2[0][0] + mat1[0][1] * mat2[1][0]) % 1000000,
                (mat1[0][0] * mat2[0][1] + mat1[0][1] * mat2[1][1]) % 1000000},
                {(mat1[1][0] * mat2[0][0] + mat1[1][1] * mat2[1][0]) % 1000000,
                        (mat1[1][0] * mat2[0][1] + mat1[1][1] * mat2[1][1]) % 1000000}};
    }

    public static long[][] matrix_pow(BigInteger exponent, long[][] matrix){
        if (exponent.compareTo(new BigInteger("1")) > 0){
            long[][] res = matrix_pow(exponent.divide(new BigInteger("2")), matrix);
            if (exponent.mod(new BigInteger("2")).equals(new BigInteger("1"))){
                return matrix_multiply(matrix, matrix_multiply(res, res));
            }
            return matrix_multiply(res, res);
        }
        return matrix;
    }

    public static long fibonacci(BigInteger n){
        long [][] matrix = {{1, 1}, {1, 0}};
        return matrix_pow(n, matrix)[1][0];
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        bw.write(Long.toString(fibonacci(new BigInteger(br.readLine()))));

        bw.flush();
        bw.close();
        br.close();
    }
}

File: 2023/04.10/서지훈_백준_2749_피보나치수3.kt
import java.io.*
import java.math.BigInteger

fun matrixMultiply(mat1: Array<LongArray>, mat2: Array<LongArray>): Array<LongArray> {
    return arrayOf(
        longArrayOf(
            (mat1[0][0] * mat2[0][0] + mat1[0][1] * mat2[1][0]) % 1000000,
            (mat1[0][0] * mat2[0][1] + mat1[0][1] * mat2[1][1]) % 1000000
        ), longArrayOf(
            (mat1[1][0] * mat2[0][0] + mat1[1][1] * mat2[1][0]) % 1000000,
            (mat1[1][0] * mat2[0][1] + mat1[1][1] * mat2[1][1]) % 1000000
        )
    )
}

fun matrixPow(exponent: BigInteger, matrix: Array<LongArray>): Array<LongArray> {
    if (exponent.compareTo(BigInteger("1")) > 0) {
        val res = matrixPow(exponent.divide(BigInteger("2")), matrix)
        return if (exponent.mod(BigInteger("2")) == BigInteger("1")) {
            matrixMultiply(matrix, matrixMultiply(res, res))
        } else matrixMultiply(res, res)
    }
    return matrix
}

fun fibonacci(n: BigInteger): Long {
    val matrix = arrayOf(longArrayOf(1, 1), longArrayOf(1, 0))
    return matrixPow(n, matrix)[1][0]
}

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    bw.write(fibonacci(BigInteger(br.readLine())).toString())

    bw.flush()
    bw.close()
    br.close()
}